java/ql/src/semmle/code/java/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1565,25 +1613,41 @@ private predicate paramFlowsThrough(ParameterNode p, CallContextCall cc, Configu
   )
 }
 
-/**
- * Holds if data may flow from `mid` to an argument of `methodcall`,
- * through a called method `m`, and back out through a return statement in
- * `m`. The context `cc` is restored to its value prior to entering `m`.
- */
 pragma[noinline]
-private predicate flowThroughMethod(PathNodeMid mid, Call methodcall, CallContext cc) {
+private predicate flowThroughCallable0(
+  DataFlowCall call, PathNodeMid mid, ReturnKind kind, CallContext cc
+) {
   exists(ParameterNode p, CallContext innercc |
-    flowIntoCallable(mid, p, cc, innercc, methodcall) and
-    paramFlowsThrough(p, innercc, unbind(mid.getConfiguration())) and
-    not parameterValueFlowsThrough(p) and
+    flowIntoCallable(mid, p, cc, innercc, call) and
+    paramFlowsThrough(p, kind, innercc, unbind(mid.getConfiguration())) and
+    not parameterValueFlowsThrough(p, kind, innercc) and
     mid.getAp() instanceof AccessPathNil
   )
 }
 
-private predicate valueFlowThroughMethod(PathNodeMid mid, Call methodcall, CallContext cc) {
-  exists(ParameterNode p |
-    flowIntoCallable(mid, p, cc, _, methodcall) and
-    parameterValueFlowsThrough(p)
+/**
+ * Holds if data may flow from `mid` through a callable to the node `out`.
+ * The context `cc` is restored to its value prior to entering the callable.
+ */
+pragma[noinline]
+private predicate flowThroughCallable(PathNodeMid mid, OutNode out, CallContext cc) {
+  exists(DataFlowCall call, ReturnKind kind | flowThroughCallable0(call, mid, kind, cc) |
+    out = getAnOutNode(call, kind)
+  )
+}
+
+pragma[noinline]
+private predicate valueFlowThroughCallable0(
+  DataFlowCall call, PathNodeMid mid, ReturnKind kind, CallContext cc
+) {
+  exists(ParameterNode p, CallContext innercc | flowIntoCallable(mid, p, cc, innercc, call) |
+    parameterValueFlowsThrough(p, kind, innercc)
+  )
+}
+
+private predicate valueFlowThroughCallable(PathNodeMid mid, OutNode out, CallContext cc) {
+  exists(ReturnKind kind | valueFlowThroughCallable0(out.getCall(), mid, kind, cc) |
```

Comment:
`valueFlowThroughCallable0(call, mid, kind, cc) and out = getAnOutNode(call, kind)` as above is simpler.